Small macro fixes

ceolin opened the pull request:
Simplify macro to get current cpu.

Commit:
xtensa: irq: Remove CURR_CPU

Use _current_cpu->id instead of CURR_CPU since that works
with/without multicore and contains additional checks.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

rebase xtensa

Commit:
pm: Remove CURRENT_CPU macro

Just use _current_cpu that works with/without multicore.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

andyross reviewed (commented):
Can you fix these to use _current_cpu (from kernel_struct.h) instead?  That macro includes an assertion that's proven really important for catching mistakes.  In general it's not legal to ask "what CPU am I running on?", because you might be preempted and migrated before you can use the answer.  _current_cpu checks that you're in a context (e.g. interrupts locked, ISR, pinned thread) where that's impossible.

Commit:
arc: irq: Remove CURR_CPU macro

Use _current_cpu->id insteado of CURR_CPU because that works
with/without multicore and contains additional safety checks.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

ceolin commented:
> Can you fix these to use _current_cpu (from kernel_struct.h) instead? That macro includes an assertion that's proven really important for catching mistakes. In general it's not legal to ask "what CPU am I running on?", because you might be preempted and migrated before you can use the answer. _current_cpu checks that you're in a context (e.g. interrupts locked, ISR, pinned thread) where that's impossible.

yep, we can. I don't quite remember why it was not using that to be honest.

evgeniy-paltsev commented on arch/arc/core/irq_offload.c:
```diff
@@ -36,13 +35,13 @@ static void arc_irq_offload_handler(const void *unused)
 {
-	offload_params[CURR_CPU].fn = routine;
-	offload_params[CURR_CPU].arg = parameter;
+	offload_params[_current_cpu->id].fn = routine;
+	offload_params[_current_cpu->id].arg = parameter;
```
Hmmm, now this require to be guarded by interrupt lock - otherwise we'll get assertion on SMP platform (due to _current_cpu macro).

ceolin replied:
Yep

ceolin commented on arch/arc/core/irq_offload.c:
```diff
@@ -36,13 +35,17 @@ static void arc_irq_offload_handler(const void *unused)
-	offload_params[CURR_CPU].arg = parameter;
+	unsigned int key;
+
+	key = irq_lock();
```
good point !

ceolin commented:
@evgeniy-paltsev would you mind take another look please ?

ceolin commented:
@evgeniy-paltsev ping
